xds/src/main/java/io/grpc/xds/XdsClient.java

```diff
@@ -81,9 +184,67 @@ ConfigUpdate build() {
    * configurations for traffic control such as drop overloads, inter-cluster load balancing
    * policy and etc.
    */
-  // TODO(zdapeng): content TBD.
   static final class EndpointUpdate {
+    private final String clusterName;
+    private final Map<Locality, LocalityLbEndpoints> localityLbEndpointsMap;
+    private final List<DropOverload> dropPolicies;
+
+    private EndpointUpdate(
+        String clusterName, Map<Locality,
+        LocalityLbEndpoints> localityLbEndpoints,
+        List<DropOverload> dropPolicies) {
+      this.clusterName = clusterName;
+      this.localityLbEndpointsMap = localityLbEndpoints;
+      this.dropPolicies = dropPolicies;
+    }
+
+    static Builder newBuilder() {
+      return new Builder();
+    }
 
+    String getClusterName() {
+      return clusterName;
+    }
+
+    /**
+     * Returns a map of localities with endpoints load balancing information in each locality.
+     */
+    Map<Locality, LocalityLbEndpoints> getLocalityLbEndpointsMap() {
+      return Collections.unmodifiableMap(localityLbEndpointsMap);
+    }
+
+    /**
+     * Returns a list of drop policies to be applied to outgoing requests.
+     */
+    List<DropOverload> getDropPolicies() {
+      return Collections.unmodifiableList(dropPolicies);
+    }
+
+    static final class Builder {
+      private String clusterName;
+      private Map<Locality, LocalityLbEndpoints> localityLbEndpointsMap = new HashMap<>();
```

Comment:
Preserve the order of entries being added?

Reply:
Hmm... Unless there is a strong (design-specific) reason to do so, do we? If it's only the ease of testing, that's wrong, we should fix that this time while we are migrating.

Reply:
The builder does not know whether the order of items in the response data matters, so it should keep is the original order. 

As for the ease of testing random picking when you use fake random: Otherwise you don't know each entry is in which position (It's deterministic and not flaky, but you don't know how it's hashed), so fake random can not return a pre-expected entry. Who do you fix those tests?

Reply:
That specific case is subtle... One approach I can think is to make `EndpointUpdate`'s constructor (currently private) `@VisibleForTesting` when you are writing the test. So that you can construct an `EndpointUpdate` instance with ordered map just for your test.

Reply:
In general we should always keep the order. Pick-first, round-robin, WRR all depends on order. Only the special case when randomness is used the order does not matter. Builder should have no prior knowledge of how policy will process the update. We should have the flexibility if we deceived to change to use WRR instead of weighted-random-picking.

Reply:
For whatever picking policy, how data presents should not affect its behavior. I mean behavior, not deterministically which one will be picked. For example, pick-first should pick the first one of the group, the policy does not care about how elements in this group are ordered. We just need to ensure doing pick-first multiple times for the __same group__ always yields the same outcome, which is considered to be "the first". Similarly, round-robin and WRR's behavior is to go through elements with rotation, the order of how they are rotated does not matter. The order of how elements are represented is an implementation detail of the container, it should be opaque to consumers as long as the design does not specifies it. Consumer implementations relying on the how data is represented internally in provider is wrong. Here all we want is deterministic output given a deterministic input for tests. For implementation, we do not need this determinism. So using `Endpoint` constructor to craft such a deterministic input for test only does what we want.

Reply:
>We just need to ensure doing pick-first multiple times for the same group always yields the same outcome, which is considered to be "the first".

That's not "the first". The first should be the first item from the remote balancer response. Grpclb may deliberately shuffle the items and return to different clients. If all the clients just "always yields the same outcome" then the shuffling effort is in vain.

Reply:
There is some misunderstanding. "pick-first" and "round robin" are names based on implementation. From the clients' perspective, they have no knowledge about the order of elements, we cannot say its behavior is wrong if the implementation takes in [A, B, C] and shuffles to [B, C, A] and then yields B. The randomness comes from the shuffling. If we make the shuffling deterministic, we always get the same result. Here our problem is to have a deterministic test, so we eliminate the randomness in test (e.g., shuffling gives the same output as input), then we could say that our implementation is correct if given [A, B, C] yields A and given [B, C, A] yields B.

Reply:
Ok. We make the implementation more conservative by preserving the ordering of elements put into this structure.